runtime/eventring_consumer.c

```diff
@@ -0,0 +1,699 @@
+  int (*runtime_begin)(int domain_id, void *callback_data, uint64_t timestamp,
+                        ev_runtime_phase phase);
+  int (*runtime_end)(int domain_id, void *callback_data, uint64_t timestamp,
+                      ev_runtime_phase phase);
+  int (*runtime_counter)(int domain_id, void *callback_data,
+                          uint64_t timestamp, ev_runtime_counter counter,
+                          uint64_t val);
+  int (*alloc)(int domain_id, void *callback_data, uint64_t timestamp,
+                uint64_t *sz);
+  int (*lifecycle)(int domain_id, void *callback_data, int64_t timestamp,
+                    ev_lifecycle lifecycle, int64_t data);
+  int (*lost_events)(int domain_id, void *callback_data, int lost_words);
+};
+
+/* C-API for reading from an eventring */
+
+eventring_error
+caml_eventring_create_cursor(const char_os* eventring_path, int pid,
+                             struct caml_eventring_cursor **cursor_res) {
+  int ret;
+
+#ifndef _WIN32
+  int ring_fd;
+  struct stat tmp_stat;
+#endif
+
+  struct caml_eventring_cursor *cursor =
+      caml_stat_alloc_noexc(sizeof(struct caml_eventring_cursor));
+  char_os *eventring_loc;
+
+  if (cursor == NULL) {
+    return E_ALLOC_FAIL;
+  }
+
+  eventring_loc = caml_stat_alloc_noexc(RING_FILE_NAME_MAX_LEN);
+
+  if (eventring_loc == NULL) {
+    return E_ALLOC_FAIL;
+  }
+
+  /* If pid < 0 then we create a cursor for the current process */
+  if (pid < 0) {
+    eventring_loc = caml_eventring_current_location();
+
+    if( eventring_loc == NULL ) {
+      return E_NO_CURRENT_RING;
+    }
+  } else {
+  /* In this case we are reading the ring for a different process */
+    if (eventring_path) {
+      char* path_u8 = caml_stat_strdup_of_os(eventring_path);
+      ret = snprintf_os(eventring_loc, RING_FILE_NAME_MAX_LEN,
+                      T("%s/%d.eventring"), path_u8, pid);
+      caml_stat_free(path_u8);
+    } else {
+      ret =
+          snprintf_os(eventring_loc, RING_FILE_NAME_MAX_LEN,
+                      T("%d.eventring"), pid);
+    }
+
+    if (ret < 0) {
+      caml_stat_free(cursor);
+      caml_stat_free(eventring_loc);
+      return E_PATH_FAILURE;
+    }
+  }
+
+#ifdef _WIN32
+  cursor->ring_file_handle = CreateFile(
+    eventring_loc,
+    GENERIC_READ | GENERIC_WRITE,
+    FILE_SHARE_READ | FILE_SHARE_WRITE,
+    NULL,
+    OPEN_EXISTING,
+    FILE_ATTRIBUTE_NORMAL,
+    NULL
+  );
+
+  if (cursor->ring_file_handle == INVALID_HANDLE_VALUE) {
+    caml_stat_free(cursor);
+    caml_stat_free(eventring_loc);
+    return E_OPEN_FAILURE;
+  }
+
+  cursor->ring_handle = CreateFileMapping(
+    cursor->ring_file_handle,
+    NULL,
+    PAGE_READWRITE,
+    0,
+    0,
+    NULL
+  );
+
+  if (cursor->ring_handle == INVALID_HANDLE_VALUE) {
+    caml_stat_free(cursor);
+    caml_stat_free(eventring_loc);
+    return E_MAP_FAILURE;
+  }
+
+  cursor->metadata = MapViewOfFile(
+    cursor->ring_handle,
+    FILE_MAP_ALL_ACCESS,
+    0,
+    0,
+    0
+  );
```

Comment:
There is a fair amount of redundancy between the code to initialize the "ring handle" and "metadata" in `eventring_consumer.c:caml_eventring_create_cursor` and the code in `eventring.c:stw_create_eventring`, would it be possible to factorize the two?

I guess it's the same for `stw_teardown_eventring` and `caml_eventring_free_cursor` below.

Reply:
So they originally did co-exist and share code but I separated them out for a couple of reasons:

1. While the producer and consumer code is very similar now, I anticipate the consumer code might get more complicated as we want to support reading eventrings from slightly older versions of OCaml. The metadata includes a version to allow for that in future.

2. There were many bugs whereby I'd accidentally use some of the global eventring _producer_ state when I should have been using the cursor's state instead.